mantidproject/mantid
# Make MaxRecursionDepth property accessible at CutMD

Users expect the PnBins properties of CutMD to exactly define the bin (in the case of MDHistoWorkspace output) or box (in the case of MDEventWorkspace output) structure of the output workspace. This was not always the case as if a box contains over a certain number of events it is split into more boxes. This behaviour can be suppressed by setting MaxRecursionDepth=1 in SliceMD, but not in its parent algorithm CutMD.

I have made MaxRecursionDepth accessible at CutMD and set its default value to be 1. A new unit test covers the expected behaviour.

Release notes (f0aa625) and user documentation for CutMD have been updated.

**To test:**
- Get the data to test with from //olympic/babylon5/Scratch/Matthew Jones/Fe_data.
- In the script below, replace the path with the correct path to the data, then run it.
- The cuts produce an MDEventWorkspace and an MDHistoWorkspace which should look identical in the plots. You could also view them in the Slice Viewer; the box structure of the MDEventWorkspace should look exactly the same as the bin structure of the MDHistoWorkspace. (Make sure you are viewing them with the same normalisation.)

``` python
import numpy as np

# Create arrays of run numbers and corresponding values of psi
run_numbers = range(15052, 15062)
psi_array = np.arange(0.0, 20.0, 2)

input_runs = ['/path/to/Fe_data/map' + str(run_number) + '_ei400.nxspe' for run_number in run_numbers]

md_ws = CreateMD(input_runs, Emode='Direct', Alatt=[2.87, 2.87, 2.87], Angdeg=[90, 90, 90], u=[1, 0, 0, ], v=[0, 1, 0], Psi=psi_array, EFix=400.0, InPlace=True)

from mantid.api import Projection
# Create an identity projection
proj_id = Projection([1,1,0], [-1,1,0], [0,0,1])
proj_id.setType(0, 'r')
proj_id.setType(1, 'r')
proj_id.setType(2, 'r')
ws_id = proj_id.createWorkspace()

out_md_1d_pix = CutMD(md_ws, Projection=ws_id, PBins=([-3,0.05,3],[-1.1,-0.9],[-0.1,0.1],[130,150]), NoPix=False)
out_md_1d_nopix = CutMD(md_ws, Projection=ws_id, PBins=([-3,0.05,3],[-1.1,-0.9],[-0.1,0.1],[130,150]), NoPix=True)

plotMD(out_md_1d_pix,error_bars=True,normalization=2)
plotMD(out_md_1d_nopix,error_bars=True,normalization=2)
```



Fixes #15831.



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
